music21/musicxml/m21ToXml.py

```diff
@@ -2541,7 +2556,34 @@ def instrumentSetup(self):
         if firstInstId in instIdList or firstInstId is None:  # must have unique ids
             self.firstInstrumentObject.partIdRandomize()  # set new random id
 
+        if self.parent is not None:
+            groups = self.parent.joinedGroups
+        else:
+            groups = []
+        for joined_group in groups:
+            if joined_group.isFirst(self.stream):
+                # need to insert instruments from subsequent staffs
+                for subseq_staff in joined_group[1:]:
+                    other_insts = subseq_staff.getInstruments(returnDefault=True, recurse=True)
+                    self.instrumentStream += other_insts
+                    deduplicate(self.instrumentStream, inPlace=True)
+                    # Place a reference to this instrument stream in a place
+                    # where subsequent staffs entering this method will find and use it
+                    if self.parent:
+                        self.parent.instrumentsByStream[id(subseq_staff)] = self.instrumentStream
+            elif self.stream in joined_group:
+                # This stream was already (or will be) processed
+                # UNLESS there is a spaghetti case where
+                # this stream is second in groupB, but first in groupA,
+                # but PartStaffExporterMixin guarantees that won't happen
+                return
+
```

Comment:
Call out to a separate routine to explain a little more what's happening and keep the general case simpler.

Reply:
thanks for tackling the refactor!